File: backend/src/main/java/com/dreampath/service/chatbot/RagSearchService.java
package com.dreampath.service.chatbot;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;

@Service
@RequiredArgsConstructor
public class RagSearchService {

    @Value("${pinecone.api.key}")
    private String pineconeApiKey;

    @Value("${pinecone.index}")
    private String pineconeIndex;

    @Value("${pinecone.host}")
    private String pineconeHost;

    private final OkHttpClient client = new OkHttpClient();

    public JSONArray search(float[] vector) {
        // Pinecone이 설정되지 않았으면 빈 배열 반환
        if ("dummykey".equals(pineconeApiKey) || "dummy-host.pinecone.io".equals(pineconeHost)) {
            return new JSONArray();
        }

        try {
            JSONObject json = new JSONObject();
            json.put("vector", vector);
            json.put("topK", 5);
            json.put("includeMetadata", true);

            RequestBody body = RequestBody.create(
                    json.toString(),
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url("https://" + pineconeHost + "/query")
                    .addHeader("Api-Key", pineconeApiKey)
                    .post(body)
                    .build();

            Response response = client.newCall(request).execute();
            String responseBody = response.body().string();

            JSONObject obj = new JSONObject(responseBody);
            return obj.getJSONArray("matches");

        } catch (Exception e) {
            // Pinecone 검색 실패 시 빈 배열 반환 (서비스는 계속 동작)
            return new JSONArray();
        }
    }
}

File: ai-service/services/chatbot/rag/rag_search_service.py
import os
import httpx
from typing import List, Dict, Any


class RagSearchService:
    def __init__(self):
        self.pinecone_api_key = os.getenv("PINECONE_FAQ_API_KEY", "dummykey")
        self.pinecone_host = os.getenv("PINECONE_FAQ_HOST", "dummy-host.pinecone.io")

    def search(self, vector: List[float], user_type: str = "guest", top_k: int = 5) -> List[Dict[str, Any]]:
        """Pinecone에서 벡터 검색 (user_type 필터링 포함)"""
        # Pinecone이 설정되지 않았으면 빈 배열 반환
        if self.pinecone_api_key == "dummykey" or self.pinecone_host == "dummy-host.pinecone.io":
            return []

        try:
            url = f"https://{self.pinecone_host}/query"
            headers = {
                "Api-Key": self.pinecone_api_key,
                "Content-Type": "application/json"
            }
            payload = {
                "vector": vector,
                "topK": top_k,
                "includeMetadata": True,
                "filter": {
                    "user_type": {"$in": [user_type, "both"]}
                }
            }

            with httpx.Client() as client:
                response = client.post(url, json=payload, headers=headers, timeout=30.0)
                response.raise_for_status()
                data = response.json()
                return data.get("matches", [])

        except Exception as e:
            # Pinecone 검색 실패 시 빈 배열 반환 (서비스는 계속 동작)
            print(f"Pinecone 검색 실패: {str(e)}")
            return []
